File: api/app/services/scanner.py
import fnmatch
import subprocess
import tempfile
from pathlib import Path
from typing import List

from app.parsers import dependency_parsers


def clone_repository(repository_url: str) -> Path:
    temp_dir = Path(tempfile.mkdtemp())
    subprocess.run(
        ["git", "clone", "--depth=1", repository_url, str(temp_dir)], check=True
    )

    return temp_dir


def extract_dependencies(repository_path: Path):
    all_dependencies: List[tuple[str, str, str]] = []

    for path in repository_path.rglob("*"):
        for pattern, parser, ecosystem in dependency_parsers:
            if fnmatch.fnmatch(path.name, pattern):
                try:
                    dependencies = parser(path)
                    all_dependencies.extend(dependencies)
                except Exception as e:
                    print(f"Failed to parse {path.name}: {e}")

    return all_dependencies


def get_repository_dependencies(repository_url: str) -> List[tuple[str, str, str]]:
    repository_path = None
    try:
        repository_path = clone_repository(repository_url)
        dependencies = extract_dependencies(repository_path)
        return dependencies
    finally:
        if repository_path and repository_path.exists():
            subprocess.run(["rm", "-rf", str(repository_path)])

File: background-worker/internal/scanner/scanner.go
package scanner

import (
	"context"
	"fmt"
	"log"
	"os"
	"os/exec"

	"io/fs"
	"path/filepath"

	"github.com/oss-health/background-worker/internal/parsers"
	"github.com/oss-health/background-worker/internal/repository"
)

func RunDailyScan(ctx context.Context, day int, hour int) {
	repositories, err := repository.GetRepositoriesForDay(ctx, day, hour)

	if err != nil {
		log.Printf("Error fetching repositories: %v", err)
	}

	for _, repository := range repositories {
		go processRepository(ctx, repository)
	}
}

func processRepository(ctx context.Context, repo repository.Repository) {
	tempDir, err := cloneRepository(repo.URL)

	if err != nil {
		log.Printf("Failed to process %s: %v", repo.URL, err)
		repository.MarkFailed(ctx, repo.ID, err.Error())
	} else {
		repository.MarkScanned(ctx, repo.ID)
	}

	dependencies, err := extractDependencies(tempDir)

	if err != nil {
		log.Printf("Could not get dependencies for %s: %v", repo.URL, err)
	}
	log.Print(len(dependencies))

	if err := os.RemoveAll(tempDir); err != nil {
		log.Printf("failed to remove %s: %v", tempDir, err)
	}
}

func cloneRepository(url string) (string, error) {
	tempDir, err := os.MkdirTemp("", "repo-clone-*")

	if err != nil {
		return "", fmt.Errorf("failed to create temp dir: %w", err)
	}

	cmd := exec.Command("git", "clone", "--depth", "1", url, tempDir)
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr

	if err := cmd.Run(); err != nil {
		if err := os.RemoveAll(tempDir); err != nil {
			return "", fmt.Errorf("failed to remove %s: %w", tempDir, err)
		}

		return "", fmt.Errorf("git clone failed: %w", err)
	}

	return tempDir, nil

}

func extractDependencies(repositoryPath string) ([]parsers.Dependency, error) {
	var allDependencies []parsers.Dependency

	err := filepath.WalkDir(repositoryPath, func(path string, d fs.DirEntry, err error) error {
		if err != nil {
			return nil
		}

		if d.IsDir() {
			return nil
		}

		filename := filepath.Base(path)

		for _, parser := range parsers.DependencyParsers {
			match, err := filepath.Match(parser.Pattern, filename)
			if err != nil {
				continue
			}
			if !match {
				match, _ = filepath.Match(parser.Pattern, path)
			}
			if match {
				deps, err := parser.Parse(path)
				if err != nil {
					fmt.Printf("Failed to parse %s: %v\n", path, err)
					continue
				}
				allDependencies = append(allDependencies, deps...)
			}
		}

		return nil
	})

	if err != nil {
		return nil, err
	}

	return allDependencies, nil
}
